js_modules/dagster-ui/packages/ui-core/src/hooks/useWrappedQuery.tsx

```diff
@@ -0,0 +1,32 @@
+import {ApolloQueryResult, OperationVariables, QueryOptions, useApolloClient} from '@apollo/client';
+import {useMemo} from 'react';
+
+import {wrapPromise} from '../utils/wrapPromise';
+
+const defaultPostProcess = async <TQuery, TResult = any>(
+  result: ApolloQueryResult<TQuery>,
+): Promise<TResult> => {
+  return result as any;
+};
+
+export function useWrappedQuery<
+  TQuery,
+  TVariables extends OperationVariables = OperationVariables,
+  TResult = ApolloQueryResult<TQuery>,
+>(
+  options: QueryOptions<TVariables, TQuery>,
+  postProcess: (result: ApolloQueryResult<TQuery>) => Promise<TResult> = defaultPostProcess as any,
+) {
+  const client = useApolloClient();
+  return useMemo(() => {
+    const promise = new Promise<TResult>((resolve, reject) => {
+      client
+        .query<TQuery, TVariables>(options)
+        .then((result) => postProcess(result))
+        .then((processed) => resolve(processed))
+        .catch((error) => reject(error));
+    });
+
+    return wrapPromise<TResult>(promise);
+  }, [client, options, postProcess]);
+}
```

Comment:
Ahh man -- this all looks good code-wise, but I'm not wild about having to properly use these higher level helpers and utils to avoid Suspense introducing waterfalls where we've never had to worry about waterfalls before. Is all this worth it for page load tracking?

One thing is that I'd be in favor of renaming this to explain what it does not how it does it. If the goal is to avoid suspense blocking on the query result maybe we call this `useUnsuspendedQuery`